### ISSUE TITLE
RDFizer is done sequentially

### ISSUE THREAD

**Issue by xmolero** [ALIADA/aliada-tool#104]
This is a problem because when two users or more did the test in the tool, the second one have to wait the first check conversion.

The RDizer must have a REST interface that can be invoked as many times as you want, in parallel.

Could you try with you users, @agazzarini ?

![screenshot - 30_04_2015 9_38_14](https://cloud.githubusercontent.com/assets/10447860/7408392/b7653246-ef1c-11e4-97a3-868562f6e86a.png)
![screenshot - 30_04_2015 9_38_31](https://cloud.githubusercontent.com/assets/10447860/7408393/b7654d80-ef1c-11e4-8660-90652c9491c1.png)

**Comment by agazzarini** [comment]
I found several points where HTTP timeout wasn't set. Now it is set but the question you raised has little to do with RDFizer...it is a matter of the ALIADA architecture. 

The problem is not in the RDFizer itself, but resides in the outbound HTTP connections the RDFizer is making towards external resources (i.e. SPARQL endpoint). Those are I/O connections and therefore are blocking the thread that is using them. 

So basically I guess this is what happens: the first job occupies the available slots (for communicating with Virtuoso) and the second job, even if it is accepted, is queued waiting for available slots.

I tried to "comment out" each call to Virtuoso (replacing with some dummy code) and I'm able to run as many jobs as I want.

I don't know exactly how to deal with this problem (assuming this is a problem because we have to remember that ALIADA isn't a realtime application where users execute several conversion in a given moment) 

@idoiamurua any thoughts?

**Comment by idoiamurua** [comment]
I am examining Virtuoso configuration file trying to find some parameter that can improve the server performance, and therefore the time to answer to the SPARQL queries.

**Comment by agazzarini** [comment]
@idoiamurua keep in mind that (at least this is valid for me), my 
environment is not so real: a huge loss of time in I/O ops is expected, 
as Virtuoso is a remote system. In an ideal situation, I suppose ALIADA 
and Virtuoso should be on the same network

**Comment by idoiamurua** [comment]
Yes. But as Xabier comments it also happens in aliada.scnabit.net machine, where the ALIADA tool and the Virtuoso server are installed togeteher.

**Comment by idoiamurua** [comment]
I have just changed NumberOfBuffers and MaxDirtyBuffers parameter values in "virtuoso.ini" file and restarted Virtuoso in aliada.scanbit.net. 

I have put the values recommended in http://virtuoso.openlinksw.com/dataspace/doc/dav/wiki/Main/VirtRDFPerformanceTuning, and considering that the machine has 8 GB of RAM.

@xmolero Do you notice any difference?

**Comment by xmolero** [comment]
No, because each conversion also increases the used memory.

![screenshot - 30_04_2015 14_15_54](https://cloud.githubusercontent.com/assets/10447860/7412514/f4f2fe82-ef43-11e4-9409-945beb748c42.png)

**Comment by agazzarini** [comment]
Hi Xabier,
sorry but that means nothing. The memory is something that has to be 
used so that screenshot could mean everything.
Could you please explain us what are you doing? Something more about 
your context? Things like:

- are you running the latest version?
- are you able to complete at least one job?
- how many records in that job?
- are you getting OOM?
- what the log is saying?

Thanks
Andrea

**Comment by idoiamurua** [comment]
Next Monday I will try to install the latest and stable version of Virtuoso 7.2, where they have fixed memory leaks.

**Comment by xmolero** [comment]
That is! The memory leaks are the problem and when the conversion run in the server increase the used memory but then not released the memory.

I am running the latest version. The conversion isn´t possible with 10000 records and the process goes slowly.

**Comment by idoiamurua** [comment]
I have just nstalled Virtuoso 07.20.3212 in aliada.scanbit.net machine.

**Comment by xmolero** [comment]
It doesn´t fix the problem, the server increase the used memory and not release the memory.

**Comment by idoiamurua** [comment]
The RDFizer is giving the following error in "/var/log/tomcat/rdfizer.log" file:

```
2015-05-04 15:30:08,495 ERROR <RDF-IZER-00050> : RDF Store failure has been detected. See below for further details.
com.hp.hpl.jena.query.QueryParseException: Encountered " "<" "< "" at line 100, column 113.
Was expecting one of:
    <IRIref> ...
    <PNAME_NS> ...
    <PNAME_LN> ...
    <VAR1> ...
    <VAR2> ...
    "a" ...
    
	at com.hp.hpl.jena.sparql.lang.ParserARQUpdate._parse(ParserARQUpdate.java:67)
	at com.hp.hpl.jena.sparql.lang.ParserARQUpdate.parse$(ParserARQUpdate.java:40)
	at com.hp.hpl.jena.sparql.lang.UpdateParser.parse(UpdateParser.java:42)
	at com.hp.hpl.jena.update.UpdateFactory.make(UpdateFactory.java:90)
	at com.hp.hpl.jena.update.UpdateFactory.create(UpdateFactory.java:81)
	at com.hp.hpl.jena.update.UpdateFactory.create(UpdateFactory.java:59)
	at com.hp.hpl.jena.update.UpdateFactory.create(UpdateFactory.java:49)
	at eu.aliada.shared.rdfstore.RDFStoreDAO.executeInsert(RDFStoreDAO.java:184)
	at eu.aliada.rdfizer.pipeline.processors.RESTSparqlInsertProcessor.process(RESTSparqlInsertProcessor.java:69)
	at org.apache.camel.processor.DelegateSyncProcessor.process(DelegateSyncProcessor.java:63)
	at org.apache.camel.management.InstrumentationProcessor.process(InstrumentationProcessor.java:72)
	at org.apache.camel.processor.interceptor.TraceInterceptor.process(TraceInterceptor.java:163)
	at org.apache.camel.processor.RedeliveryErrorHandler.process(RedeliveryErrorHandler.java:398)
	at org.apache.camel.processor.CamelInternalProcessor.process(CamelInternalProcessor.java:191)
	at org.apache.camel.processor.CamelInternalProcessor.process(CamelInternalProcessor.java:191)
	at org.apache.camel.component.seda.SedaConsumer.sendToConsumers(SedaConsumer.java:291)
	at org.apache.camel.component.seda.SedaConsumer.doRun(SedaConsumer.java:200)
	at org.apache.camel.component.seda.SedaConsumer.run(SedaConsumer.java:147)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

```

**Comment by idoiamurua** [comment]
I have just changed the LOG level of the RDFizer from INFO to DEBUG to see which triples give that error.

**Comment by agazzarini** [comment]
Hi Idoia, I'm not in the office today. What kind of format are you trying?

**Comment by idoiamurua** [comment]
MARCXML

**Comment by agazzarini** [comment]
Ok, since @emiliano-cammilletti  is already working on that, could you please open a new issue and send him the file you were using?

**Comment by agazzarini** [comment]
Update on this: I need some more time because although the memory issue has been mitigated the XPATH usage is the main responsible of the performance degrade

**Comment by xmolero** [comment]
![memory-virtuosoandtomcat](https://cloud.githubusercontent.com/assets/10447860/7489234/ce0c3b54-f3d4-11e4-8d7e-a43f13aa2bb3.png)

Virtuoso doesn´t change the memory, definitely the problem is in Tomcat.

**Comment by idoiamurua** [comment]
@emiliano-cammilletti 
I found the bug that gives the error when inserting the triples in the RDFizer. File "marcxml.frbr.n3.vm" contains the following:
```
"#ecrm(' P14i_performed')"
```
which should be (remove blank space before the "P"):
```
"#ecrm('P14i_performed')"
```
which it was generating the following code in a triple:
```
<http://erlangen-crm.org/current/ P14i_performed>
```

**Comment by agazzarini** [comment]
Issue has been fixed with the last commit. The current implementation attach on the validation channel (which was synchronous) a set of (configurable) consumers, therefore allowing more than one "job" to be validated and passed through the pipeline. 

However, having said that, it's important to underline what I already told during the confcall: each job is a massive task that consumes a lot of resources so running several jobs in parallel, although possible, it is considered a bad pratice as it probably would compromise the overall scalability of the system.

In addition, ALIADA hasn't been implemented in a SAAS "perspective" so several parallel requests constitute a non-frequent scenario. Instead of dividing a big input dataset in several small chunks, it is strongly recommended to give the whole file in input.
